server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/fileformats/data/StarTreeDataWriter.java

```diff
@@ -0,0 +1,110 @@
+
+    /**
+     * Writes the star-tree data structure.
+     *
+     * @param indexOutput the IndexOutput to write the star-tree data
+     * @param rootNode    the root node of the star-tree
+     * @param numNodes    the total number of nodes in the star-tree
+     * @param name        the name of the star-tree field
+     * @return the total size in bytes of the serialized star-tree data
+     * @throws IOException if an I/O error occurs while writing the star-tree data
+     */
+    public static long writeStarTree(IndexOutput indexOutput, InMemoryTreeNode rootNode, int numNodes, String name) throws IOException {
+        long totalSizeInBytes = (long) numNodes * SERIALIZABLE_DATA_SIZE_IN_BYTES;
+
+        logger.debug("Star tree data size in bytes : {} for star-tree field {}", totalSizeInBytes, name);
+
+        writeStarTreeNodes(indexOutput, rootNode);
+        return totalSizeInBytes;
+    }
+
+    /**
+     * Writes the star-tree nodes in a breadth-first order.
+     *
+     * @param output   the IndexOutput to write the nodes
+     * @param rootNode the root node of the star-tree
+     * @throws IOException if an I/O error occurs while writing the nodes
+     */
+    private static void writeStarTreeNodes(IndexOutput output, InMemoryTreeNode rootNode) throws IOException {
+        Queue<InMemoryTreeNode> queue = new LinkedList<>();
+        queue.add(rootNode);
+
+        int currentNodeId = 0;
+        while (!queue.isEmpty()) {
+            InMemoryTreeNode node = queue.remove();
+
+            if (node.children == null || node.children.isEmpty()) {
+                writeStarTreeNode(output, node, ALL, ALL);
+            } else {
+
+                // Sort all children nodes based on dimension value
+                List<InMemoryTreeNode> sortedChildren = new ArrayList<>(node.children.values());
+                sortedChildren.sort(
+                    Comparator.comparingInt(InMemoryTreeNode::getNodeType).thenComparingLong(InMemoryTreeNode::getDimensionValue)
+                );
+
+                int firstChildId = currentNodeId + queue.size() + 1;
+                int lastChildId = firstChildId + sortedChildren.size() - 1;
+                writeStarTreeNode(output, node, firstChildId, lastChildId);
+
+                queue.addAll(sortedChildren);
+            }
+
+            currentNodeId++;
+        }
+    }
+
+    /**
+     * Writes a single star-tree node
+     *
+     * @param output       the IndexOutput to write the node
+     * @param node         the star tree node to write
+     * @param firstChildId the ID of the first child node
+     * @param lastChildId  the ID of the last child node
+     * @throws IOException if an I/O error occurs while writing the node
+     */
+    private static void writeStarTreeNode(IndexOutput output, InMemoryTreeNode node, int firstChildId, int lastChildId) throws IOException {
+        output.writeInt(node.dimensionId);
```

Comment:
Why can we not use vInt/Vlong? What are the challenged with supporting variable length nodes?

Reply:
We use random access to directly read a particular attribute of a node because we store the node with fixed length data types. With variable ints/ longs, its not straightforward to directly perform random access. It would be a tricky implementation, with edge cases and might involve us storing the length of each node attribute value, which will have its own storage costs.